kyuubi-hive-beeline/src/main/java/org/apache/hive/beeline/hs2connection/KyuubiConfFileParser.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hive.beeline.hs2connection;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements HS2ConnectionFileParser for the user-specific connection configuration file
+ * named kyuubi-defaults.conf. The class looks for this file in ${KYUUBI_CONF_DIR} in that order and
+ * uses the first file found in the above locations.
+ */
+public class KyuubiConfFileParser implements HS2ConnectionFileParser {
+
+  public static final String DEFAULT_CONNECTION_CONFIG_FILE_NAME = "kyuubi-defaults.conf";
+
+  private final List<String> locations = new ArrayList<>();
+  private static final Logger log = LoggerFactory.getLogger(KyuubiConfFileParser.class);
+
+  public KyuubiConfFileParser() {
+    if (System.getenv("KYUUBI_CONF_DIR") != null) {
+      locations.add(
+          System.getenv("KYUUBI_CONF_DIR") + File.separator + DEFAULT_CONNECTION_CONFIG_FILE_NAME);
+    }
+  }
+
+  @VisibleForTesting
+  KyuubiConfFileParser(List<String> testLocations) {
+    if (testLocations == null) {
+      return;
+    }
+    locations.addAll(testLocations);
+  }
+
+  @Override
+  public Properties getConnectionProperties() throws KyuubiConfFileParseException {
+    Properties props = new Properties();
+    String fileLocation = getFileLocation();
+    if (fileLocation == null) {
+      log.debug("Kyuubi default configuration file not found");
+      return props;
+    }
+    log.info("Using Kyuubi default configuration file at " + fileLocation);
+    File propsFile = new File(fileLocation);
+    try {
+      Map<String, String> propsMap = getPropertiesFromFile(propsFile);
+      StringBuilder hiveConf = new StringBuilder();
+      for (Entry<String, String> kv : propsMap.entrySet()) {
```

Comment:
Load all configurations? Should we use a fixed prefix like `BEELINE_CONNECTION_PROPERTY_PREFIX`?

Reply:
the intention is infer Kyuubi connection info from the existing `kyuubi-defaults.conf`, instead supporting configuring beeline specific configurations inside `kyuubi-defaults.conf`

Reply:
Currently, most of configurations in `kyuubi-defaults.conf` are server-side configurations. We can define a prefix to determine whether it is a client config, and then add them to hive_conf of jdbc url.

In addition to inferring `HIVE_CONF_PROPERTY_KEY`, we also need to infer `HOST_PROPERTY_KEY` based on server configuration.

Reply:
> We can define a prefix to determine whether it is a client config

`beeline-site.xml` takes responsibility for that, I don't think we should propagate it to `kyuubi-defaults.conf`

Why do we need `HOST_PROPERTY_KEY`?

Reply:
> beeline-site.xml takes responsibility for that, I don't think we should propagate it to kyuubi-defaults.conf

In addition to kyuubi server `host` and `port`, what other valid configurations can we infer in the current `kyuubi-defaults.conf`?

> Why do we need HOST_PROPERTY_KEY?

Pass server host and port configured in kyuubi-defaults.conf to jdbc url.